File: investment-plugin/scripts/yahoo-fetch.ts
/**
 * Yahoo Finance Data Fetcher
 *
 * Usage: npx tsx yahoo-fetch.ts <TICKER>
 * Output: JSON to stdout with quote, financials, balance sheet, cash flow,
 *         historical prices, and analyst ratings.
 *
 * Dependencies: yahoo-finance2
 */

import yahooFinance from "yahoo-finance2";

const ticker = process.argv[2];

if (!ticker) {
  console.error("Usage: npx tsx yahoo-fetch.ts <TICKER>");
  process.exit(1);
}

interface FetchResult {
  ticker: string;
  fetchDate: string;
  quote: Record<string, unknown> | null;
  financials: Record<string, unknown> | null;
  balanceSheet: Record<string, unknown> | null;
  cashFlow: Record<string, unknown> | null;
  historicalPrices: unknown[] | null;
  analystRatings: Record<string, unknown> | null;
  errors: string[];
}

async function fetchAll(symbol: string): Promise<FetchResult> {
  const result: FetchResult = {
    ticker: symbol,
    fetchDate: new Date().toISOString().split("T")[0],
    quote: null,
    financials: null,
    balanceSheet: null,
    cashFlow: null,
    historicalPrices: null,
    analystRatings: null,
    errors: [],
  };

  // Quote summary
  try {
    const quote = await yahooFinance.quoteSummary(symbol, {
      modules: [
        "price",
        "summaryDetail",
        "defaultKeyStatistics",
        "financialData",
        "earningsTrend",
        "recommendationTrend",
      ],
    });
    result.quote = {
      price: quote.price?.regularMarketPrice,
      marketCap: quote.price?.marketCap,
      currency: quote.price?.currency,
      exchange: quote.price?.exchangeName,
      name: quote.price?.longName || quote.price?.shortName,
      pe: quote.summaryDetail?.trailingPE,
      forwardPe: quote.summaryDetail?.forwardPE,
      pb: quote.defaultKeyStatistics?.priceToBook,
      ps: quote.summaryDetail?.priceToSalesTrailing12Months,
      evToEbitda: quote.defaultKeyStatistics?.enterpriseToEbitda,
      evToRevenue: quote.defaultKeyStatistics?.enterpriseToRevenue,
      dividendYield: quote.summaryDetail?.dividendYield,
      beta: quote.defaultKeyStatistics?.beta,
      fiftyTwoWeekHigh: quote.summaryDetail?.fiftyTwoWeekHigh,
      fiftyTwoWeekLow: quote.summaryDetail?.fiftyTwoWeekLow,
      avgVolume: quote.summaryDetail?.averageVolume,
      sharesOutstanding: quote.defaultKeyStatistics?.sharesOutstanding,
      floatShares: quote.defaultKeyStatistics?.floatShares,
      shortRatio: quote.defaultKeyStatistics?.shortRatio,
      targetMeanPrice: quote.financialData?.targetMeanPrice,
      targetHighPrice: quote.financialData?.targetHighPrice,
      targetLowPrice: quote.financialData?.targetLowPrice,
      numberOfAnalystOpinions:
        quote.financialData?.numberOfAnalystOpinions,
      recommendationKey: quote.financialData?.recommendationKey,
      grossMargin: quote.financialData?.grossMargins,
      operatingMargin: quote.financialData?.operatingMargins,
      profitMargin: quote.financialData?.profitMargins,
      returnOnEquity: quote.financialData?.returnOnEquity,
      returnOnAssets: quote.financialData?.returnOnAssets,
      revenueGrowth: quote.financialData?.revenueGrowth,
      earningsGrowth: quote.financialData?.earningsGrowth,
      totalDebt: quote.financialData?.totalDebt,
      totalCash: quote.financialData?.totalCash,
      debtToEquity: quote.financialData?.debtToEquity,
      currentRatio: quote.financialData?.currentRatio,
      freeCashflow: quote.financialData?.freeCashflow,
      operatingCashflow: quote.financialData?.operatingCashflow,
    };
  } catch (e) {
    result.errors.push(`quote: ${(e as Error).message}`);
  }

  // Income statement (annual + quarterly)
  try {
    const annual = await yahooFinance.fundamentalsTimeSeries(symbol, {
      period1: "2022-01-01",
      type: "annual",
      module: "financials",
    });
    const quarterly = await yahooFinance.fundamentalsTimeSeries(symbol, {
      period1: "2024-01-01",
      type: "quarterly",
      module: "financials",
    });
    result.financials = { annual, quarterly };
  } catch (e) {
    result.errors.push(`financials: ${(e as Error).message}`);
  }

  // Balance sheet
  try {
    const annual = await yahooFinance.fundamentalsTimeSeries(symbol, {
      period1: "2022-01-01",
      type: "annual",
      module: "balance-sheet",
    });
    result.balanceSheet = { annual };
  } catch (e) {
    result.errors.push(`balanceSheet: ${(e as Error).message}`);
  }

  // Cash flow
  try {
    const annual = await yahooFinance.fundamentalsTimeSeries(symbol, {
      period1: "2022-01-01",
      type: "annual",
      module: "cash-flow",
    });
    result.cashFlow = { annual };
  } catch (e) {
    result.errors.push(`cashFlow: ${(e as Error).message}`);
  }

  // Historical prices (2 years)
  try {
    const twoYearsAgo = new Date();
    twoYearsAgo.setFullYear(twoYearsAgo.getFullYear() - 2);
    const history = await yahooFinance.chart(symbol, {
      period1: twoYearsAgo.toISOString().split("T")[0],
      interval: "1wk",
    });
    result.historicalPrices = history.quotes?.map((q) => ({
      date: q.date,
      open: q.open,
      high: q.high,
      low: q.low,
      close: q.close,
      volume: q.volume,
    })) || [];
  } catch (e) {
    result.errors.push(`historicalPrices: ${(e as Error).message}`);
  }

  // Analyst ratings (from recommendationTrend)
  try {
    const rec = await yahooFinance.quoteSummary(symbol, {
      modules: ["recommendationTrend"],
    });
    result.analystRatings = {
      trend: rec.recommendationTrend?.trend,
    };
  } catch (e) {
    result.errors.push(`analystRatings: ${(e as Error).message}`);
  }

  return result;
}

fetchAll(ticker)
  .then((data) => {
    console.log(JSON.stringify(data, null, 2));
  })
  .catch((err) => {
    console.error(`Fatal error: ${err.message}`);
    process.exit(1);
  });

File: investment-plugin/skills/data-fetch/scripts/yahoo_fetch.py
#!/usr/bin/env python3
"""
Yahoo Finance Data Fetcher

Usage: python3 yahoo_fetch.py <TICKER>
Output: JSON to stdout with quote, financials, balanceSheet, cashFlow,
        historicalPrices, analystRatings, errors.

Dependencies: yfinance
"""

import json
import math
import sys
from datetime import datetime, timedelta

try:
    import yfinance as yf
except ImportError:
    print("Error: yfinance not installed. Run: pip install -r requirements.txt", file=sys.stderr)
    sys.exit(1)


def fetch_all(symbol: str) -> dict:
    result = {
        "ticker": symbol,
        "fetchDate": datetime.now().strftime("%Y-%m-%d"),
        "quote": None,
        "financials": None,
        "balanceSheet": None,
        "cashFlow": None,
        "historicalPrices": None,
        "analystRatings": None,
        "majorHolders": None,
        "institutionalHolders": None,
        "errors": [],
    }

    t = yf.Ticker(symbol)

    # Quote summary
    try:
        info = t.info
        result["quote"] = {
            "price": info.get("currentPrice") or info.get("regularMarketPrice"),
            "marketCap": info.get("marketCap"),
            "currency": info.get("currency"),
            "exchange": info.get("exchange"),
            "name": info.get("longName") or info.get("shortName"),
            "pe": info.get("trailingPE"),
            "forwardPe": info.get("forwardPE"),
            "pb": info.get("priceToBook"),
            "ps": info.get("priceToSalesTrailing12Months"),
            "evToEbitda": info.get("enterpriseToEbitda"),
            "evToRevenue": info.get("enterpriseToRevenue"),
            "dividendYield": info.get("dividendYield"),
            "beta": info.get("beta"),
            "fiftyTwoWeekHigh": info.get("fiftyTwoWeekHigh"),
            "fiftyTwoWeekLow": info.get("fiftyTwoWeekLow"),
            "avgVolume": info.get("averageVolume"),
            "sharesOutstanding": info.get("sharesOutstanding"),
            "floatShares": info.get("floatShares"),
            "shortRatio": info.get("shortRatio"),
            "targetMeanPrice": info.get("targetMeanPrice"),
            "targetHighPrice": info.get("targetHighPrice"),
            "targetLowPrice": info.get("targetLowPrice"),
            "numberOfAnalystOpinions": info.get("numberOfAnalystOpinions"),
            "recommendationKey": info.get("recommendationKey"),
            "grossMargin": info.get("grossMargins"),
            "operatingMargin": info.get("operatingMargins"),
            "profitMargin": info.get("profitMargins"),
            "returnOnEquity": info.get("returnOnEquity"),
            "returnOnAssets": info.get("returnOnAssets"),
            "revenueGrowth": info.get("revenueGrowth"),
            "earningsGrowth": info.get("earningsGrowth"),
            "totalDebt": info.get("totalDebt"),
            "totalCash": info.get("totalCash"),
            "debtToEquity": info.get("debtToEquity"),
            "currentRatio": info.get("currentRatio"),
            "freeCashflow": info.get("freeCashflow"),
            "operatingCashflow": info.get("operatingCashflow"),
        }
    except Exception as e:
        result["errors"].append(f"quote: {e}")

    # Income statement (annual + quarterly)
    try:
        annual = t.financials
        quarterly = t.quarterly_financials
        result["financials"] = {
            "annual": _df_to_records(annual),
            "quarterly": _df_to_records(quarterly),
        }
    except Exception as e:
        result["errors"].append(f"financials: {e}")

    # Balance sheet
    try:
        annual = t.balance_sheet
        result["balanceSheet"] = {
            "annual": _df_to_records(annual),
        }
    except Exception as e:
        result["errors"].append(f"balanceSheet: {e}")

    # Cash flow
    try:
        annual = t.cashflow
        result["cashFlow"] = {
            "annual": _df_to_records(annual),
        }
    except Exception as e:
        result["errors"].append(f"cashFlow: {e}")

    # Historical prices (2 years, weekly)
    try:
        two_years_ago = datetime.now() - timedelta(days=730)
        hist = t.history(start=two_years_ago.strftime("%Y-%m-%d"), interval="1wk")
        result["historicalPrices"] = [
            {
                "date": idx.strftime("%Y-%m-%d"),
                "open": round(row["Open"], 4),
                "high": round(row["High"], 4),
                "low": round(row["Low"], 4),
                "close": round(row["Close"], 4),
                "volume": int(row["Volume"]),
            }
            for idx, row in hist.iterrows()
        ]
    except Exception as e:
        result["errors"].append(f"historicalPrices: {e}")

    # Analyst ratings
    try:
        rec = t.recommendations
        if rec is not None and not rec.empty:
            result["analystRatings"] = {
                "trend": _df_to_records(rec.tail(10)),
            }
        else:
            result["analystRatings"] = {"trend": []}
    except Exception as e:
        result["errors"].append(f"analystRatings: {e}")

    # Major holders (institutional/insider ownership %)
    try:
        major = t.major_holders
        result["majorHolders"] = _df_to_records(major) if major is not None and not major.empty else []
    except Exception as e:
        result["errors"].append(f"majorHolders: {e}")

    # Top institutional holders
    try:
        inst = t.institutional_holders
        result["institutionalHolders"] = _df_to_records(inst.head(10)) if inst is not None and not inst.empty else []
    except Exception as e:
        result["errors"].append(f"institutionalHolders: {e}")

    return result


def _df_to_records(df) -> list:
    """Convert a pandas DataFrame to a list of dicts with string keys."""
    if df is None or df.empty:
        return []
    # Transpose if columns are dates (yfinance convention)
    if hasattr(df.columns[0], "strftime"):
        df = df.T
    records = []
    for idx, row in df.iterrows():
        rec = {"date": str(idx)}
        for col in df.columns:
            val = row[col]
            # Convert numpy types to Python native
            if hasattr(val, "item"):
                val = val.item()
            # NaN → None for valid JSON
            if isinstance(val, float) and math.isnan(val):
                val = None
            rec[str(col)] = val
        records.append(rec)
    return records


def main():
    if len(sys.argv) < 2:
        print("Usage: python3 yahoo_fetch.py <TICKER>", file=sys.stderr)
        sys.exit(1)

    symbol = sys.argv[1].upper()
    data = fetch_all(symbol)
    print(json.dumps(data, indent=2, default=str))


if __name__ == "__main__":
    main()
